File: packages/agent/src/retrieval/embeddings.ts
/**
 * Voyage embeddings client — TypeScript side.
 *
 * The Python indexer (apps/indexer/src/indexer/embeddings.py) embeds
 * chunks at indexing time. This module mirrors that contract on the
 * query side: at retrieval time, we embed the user's query with the
 * same model so the dot product against indexed vectors is meaningful.
 *
 * Voyage tunes the embeddings asymmetrically based on `input_type`:
 *   - indexed chunks  → input_type: "document"
 *   - search queries  → input_type: "query"
 * Get this wrong and recall drops without any error surfacing.
 *
 * Implementation notes:
 *   - Plain `fetch` (no axios — AGENTS.md § 7).
 *   - Retry with exponential backoff on 429 + 5xx + network errors.
 *     Non-retryable 4xx fails immediately.
 *   - Constructor accepts a `fetch` override so tests can inject a
 *     stub without touching the network.
 */

const VOYAGE_API_URL = "https://api.voyageai.com/v1/embeddings";
const DEFAULT_MODEL = "voyage-code-3";
const DEFAULT_TIMEOUT_MS = 30_000;
const DEFAULT_MAX_ATTEMPTS = 5;
const RETRYABLE_STATUS = new Set([429, 500, 502, 503, 504]);

export type Vector = number[];

export class EmbeddingError extends Error {
  constructor(
    message: string,
    public readonly status?: number,
  ) {
    super(message);
    this.name = "EmbeddingError";
  }
}

export type VoyageClientOptions = {
  apiKey: string;
  model?: string;
  timeoutMs?: number;
  maxAttempts?: number;
  /** Injectable for tests; defaults to global fetch. */
  fetchImpl?: typeof fetch;
};

/** Voyage embeddings client. Use {@link VoyageClient.embedQuery} on the
 *  retrieval path; {@link VoyageClient.embedDocuments} exists for symmetry. */
export class VoyageClient {
  private readonly apiKey: string;
  private readonly model: string;
  private readonly timeoutMs: number;
  private readonly maxAttempts: number;
  private readonly fetchImpl: typeof fetch;

  constructor(opts: VoyageClientOptions) {
    if (!opts.apiKey) throw new Error("VoyageClient: apiKey is required");
    this.apiKey = opts.apiKey;
    this.model = opts.model ?? DEFAULT_MODEL;
    this.timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.maxAttempts = opts.maxAttempts ?? DEFAULT_MAX_ATTEMPTS;
    this.fetchImpl = opts.fetchImpl ?? globalThis.fetch;
  }

  async embedQuery(query: string): Promise<Vector> {
    const vectors = await this.embed([query], "query");
    const first = vectors[0];
    if (!first) throw new EmbeddingError("Voyage returned no vectors for query");
    return first;
  }

  async embedDocuments(texts: string[]): Promise<Vector[]> {
    if (texts.length === 0) return [];
    return this.embed(texts, "document");
  }

  // ── internals ─────────────────────────────────────────────────────

  private async embed(input: string[], inputType: "query" | "document"): Promise<Vector[]> {
    let lastError: unknown;
    for (let attempt = 1; attempt <= this.maxAttempts; attempt++) {
      try {
        return await this.requestOnce(input, inputType);
      } catch (err) {
        lastError = err;
        if (err instanceof EmbeddingError) {
          // No status = response-shape failure; retrying won't help.
          // With status = HTTP failure; retry only if in the retry set.
          if (err.status === undefined || !RETRYABLE_STATUS.has(err.status)) throw err;
        }
        if (attempt === this.maxAttempts) break;
        await sleep(backoffMs(attempt));
      }
    }
    throw lastError instanceof Error
      ? lastError
      : new EmbeddingError(`Voyage request failed: ${String(lastError)}`);
  }

  private async requestOnce(input: string[], inputType: "query" | "document"): Promise<Vector[]> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);

    let response: Response;
    try {
      response = await this.fetchImpl(VOYAGE_API_URL, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ input, model: this.model, input_type: inputType }),
        signal: controller.signal,
      });
    } finally {
      clearTimeout(timer);
    }

    if (!response.ok) {
      const body = await safeText(response);
      throw new EmbeddingError(
        `Voyage returned ${response.status}: ${body.slice(0, 200)}`,
        response.status,
      );
    }

    const payload = (await response.json()) as {
      data?: Array<{ embedding?: unknown }>;
    };
    if (!payload.data || !Array.isArray(payload.data)) {
      throw new EmbeddingError(`Voyage response missing 'data' array`);
    }
    const vectors: Vector[] = [];
    for (const item of payload.data) {
      if (!Array.isArray(item.embedding)) {
        throw new EmbeddingError("Voyage response item missing 'embedding'");
      }
      vectors.push(item.embedding as Vector);
    }
    if (vectors.length !== input.length) {
      throw new EmbeddingError(
        `Voyage returned ${vectors.length} vectors for ${input.length} inputs`,
      );
    }
    return vectors;
  }
}

function backoffMs(attempt: number): number {
  // 500ms, 1s, 2s, 4s, 8s
  return Math.min(8_000, 500 * 2 ** (attempt - 1));
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function safeText(response: Response): Promise<string> {
  try {
    return await response.text();
  } catch {
    return response.statusText;
  }
}

File: apps/indexer/src/indexer/embeddings.py
"""Voyage embeddings adapter.

Phase 2.3 — turns chunk text into the 1024-dim vectors that go into
the `chunks.embedding` column.

Why Voyage:
  - `voyage-code-3` beats general-purpose embeddings on code retrieval
    benchmarks (see ADR-001). The choice is locked in for the project.

What this module owns:
  - One client class (`VoyageEmbedder`) with two methods:
        embed_documents(texts) -> list[Vector]    # for indexing
        embed_query(query)     -> Vector          # for search-time
    The two flavors send different `input_type` to Voyage, which
    asymmetrically tunes the two ends of the dot product.
  - Batching to Voyage's per-request limit (128 inputs).
  - Retry with exponential backoff on transient errors (429 / 5xx /
    network). Non-retryable errors (auth, bad request) raise
    immediately as `EmbeddingError`.

What this module does NOT own:
  - Persistence (writing vectors into Postgres) — separate concern.
  - Concurrency — sync API for simplicity; callers can parallelize
    at a higher level if they need to.

Testability:
  - The constructor accepts an optional `httpx.Client` so tests can
    inject one backed by `httpx.MockTransport` and avoid real HTTP.
"""

from __future__ import annotations

from typing import Literal, cast

import httpx
from tenacity import (
    RetryError,
    retry,
    retry_if_exception_type,
    stop_after_attempt,
    wait_exponential,
)

from shared.config import load_settings

# Public type alias. Vectors are plain lists of floats — keeping the type
# simple makes downstream serialization trivial.
Vector = list[float]

# Voyage API constants.
_API_URL = "https://api.voyageai.com/v1/embeddings"
_DEFAULT_MODEL = "voyage-code-3"
_BATCH_SIZE = 128  # Voyage's max inputs per request as of 2026-05
_DEFAULT_TIMEOUT_SECONDS = 30.0

# Error codes we should retry. 429 and 5xx are transient; 4xx (except
# 429) almost always means we sent something Voyage can't accept and
# retrying won't help.
_RETRYABLE_STATUS = {429, 500, 502, 503, 504}


class EmbeddingError(RuntimeError):
    """Raised when Voyage returns a non-retryable error (or retries exhausted)."""


class _RetryableHTTPStatusError(RuntimeError):
    """Internal: signals tenacity that this attempt should be retried."""


class VoyageEmbedder:
    """Synchronous Voyage embeddings client."""

    def __init__(
        self,
        api_key: str,
        *,
        model: str = _DEFAULT_MODEL,
        batch_size: int = _BATCH_SIZE,
        timeout_seconds: float = _DEFAULT_TIMEOUT_SECONDS,
        client: httpx.Client | None = None,
    ) -> None:
        if not api_key:
            raise ValueError("VoyageEmbedder requires a non-empty api_key")
        self._api_key = api_key
        self._model = model
        self._batch_size = batch_size
        self._owned_client = client is None
        self._client = client or httpx.Client(timeout=timeout_seconds)

    def close(self) -> None:
        """Release the underlying httpx.Client if we own it."""
        if self._owned_client:
            self._client.close()

    def __enter__(self) -> VoyageEmbedder:
        return self

    def __exit__(self, *_: object) -> None:
        self.close()

    # ── public ──────────────────────────────────────────────────────────

    def embed_documents(self, texts: list[str]) -> list[Vector]:
        """Embed a list of indexable chunks. Returns vectors in the same order."""
        return self._embed_batched(texts, input_type="document")

    def embed_query(self, query: str) -> Vector:
        """Embed a single search query. Returns one vector."""
        vectors = self._embed_batched([query], input_type="query")
        # _embed_batched guarantees one vector per input
        return vectors[0]

    # ── internals ───────────────────────────────────────────────────────

    def _embed_batched(
        self, texts: list[str], *, input_type: Literal["document", "query"]
    ) -> list[Vector]:
        if not texts:
            return []
        out: list[Vector] = []
        for start in range(0, len(texts), self._batch_size):
            batch = texts[start : start + self._batch_size]
            out.extend(self._embed_one_request(batch, input_type=input_type))
        return out

    def _embed_one_request(
        self, batch: list[str], *, input_type: Literal["document", "query"]
    ) -> list[Vector]:
        try:
            return self._request_with_retry(batch, input_type)
        except RetryError as err:
            # tenacity wraps the final failure; surface the underlying cause.
            cause = err.last_attempt.exception() if err.last_attempt else None
            raise EmbeddingError(f"Voyage request failed after retries: {cause}") from cause

    @retry(
        retry=retry_if_exception_type((httpx.HTTPError, _RetryableHTTPStatusError)),
        wait=wait_exponential(multiplier=0.5, min=0.5, max=8.0),
        stop=stop_after_attempt(5),
        reraise=False,
    )
    def _request_with_retry(
        self, batch: list[str], input_type: Literal["document", "query"]
    ) -> list[Vector]:
        response = self._client.post(
            _API_URL,
            headers={
                "Authorization": f"Bearer {self._api_key}",
                "Content-Type": "application/json",
            },
            json={
                "input": batch,
                "model": self._model,
                "input_type": input_type,
            },
        )

        if response.status_code in _RETRYABLE_STATUS:
            raise _RetryableHTTPStatusError(
                f"voyage returned {response.status_code}: {response.text[:200]}"
            )
        if response.status_code >= 400:
            # Non-retryable client/server error: stop immediately.
            raise EmbeddingError(f"voyage returned {response.status_code}: {response.text[:200]}")

        body = response.json()
        data = body.get("data")
        if not isinstance(data, list):
            raise EmbeddingError(f"voyage response missing 'data' array: {body!r}")

        vectors: list[Vector] = []
        for item in data:
            emb = item.get("embedding")
            if not isinstance(emb, list):
                raise EmbeddingError(f"voyage response item missing 'embedding': {item!r}")
            vectors.append(cast(Vector, emb))

        if len(vectors) != len(batch):
            raise EmbeddingError(f"voyage returned {len(vectors)} vectors for {len(batch)} inputs")
        return vectors


def voyage_embedder_from_env(**kwargs: object) -> VoyageEmbedder:
    """Construct a VoyageEmbedder using credentials from `shared.config`.

    Raises EmbeddingError if VOYAGE_API_KEY isn't configured — fail loud
    rather than silently produce empty vectors.
    """
    settings = load_settings()
    if not settings.voyage_api_key:
        raise EmbeddingError("VOYAGE_API_KEY is not set. Configure it in apps/indexer/.env.")
    return VoyageEmbedder(api_key=settings.voyage_api_key, **kwargs)  # type: ignore[arg-type]
